power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/serializer.hpp

```diff
@@ -6,16 +6,306 @@
 #define POWER_GRID_MODEL_AUXILIARY_SERIALIZTION_SERIALIZER_HPP
 
+#include "../../exception.hpp"
 #include "../../power_grid_model.hpp"
+#include "../meta_data.hpp"
+#include "../meta_data_gen.hpp"
 
 #include <nlohmann/json.hpp>
 
 #include <msgpack.hpp>
 
+#include <span>
+#include <string_view>
+
+// custom packers
+namespace msgpack {
+MSGPACK_API_VERSION_NAMESPACE(MSGPACK_DEFAULT_API_NS) {
+    namespace adaptor {
+
+    // pack name of component and attribute
+    template <class T>
+        requires(std::same_as<T, PGM_MetaComponent> || std::same_as<T, PGM_MetaAttribute>)
+    struct pack<T const*> {
+        template <typename Stream>
+        msgpack::packer<Stream>& operator()(msgpack::packer<Stream>& p, T const* const& o) const {
+            p.pack(o->name);
+            return p;
+        }
+    };
+
+    // pack double[3]
+    template <> struct pack<power_grid_model::RealValue<false>> {
+        template <typename Stream>
+        msgpack::packer<Stream>& operator()(msgpack::packer<Stream>& p,
+                                            power_grid_model::RealValue<false> const& o) const {
+            p.pack_array(3);
+            for (int8_t i = 0; i != 3; ++i) {
+                if (power_grid_model::is_nan(o(i))) {
+                    p.pack_nil();
+                } else {
+                    p.pack(o(i));
+                }
+            }
+            return p;
+        }
+    };
+
+    } // namespace adaptor
+} // MSGPACK_API_VERSION_NAMESPACE(MSGPACK_DEFAULT_API_NS)
+} // namespace msgpack
+
 namespace power_grid_model::meta_data {
 
 class Serializer {
+
+    struct ComponentBuffer {
+        MetaComponent const* component;
+        void const* data;
+        Idx size;
+    };
+
+    struct ScenarioBuffer {
+        std::vector<ComponentBuffer> component_buffers;
+    };
+
   public:
+    static constexpr std::string_view version = "1.0";
+    // top dict: version, type, is_batch, attributes, data
+    static constexpr size_t size_top_dict = 5;
+
+    // not copyable
+    Serializer(Serializer const&) = delete;
+    Serializer& operator=(Serializer const&) = delete;
+    // not movable
+    Serializer(Serializer&&) = delete;
+    Serializer& operator=(Serializer&&) = delete;
+    // destructor
+    ~Serializer() = default;
+
+    Serializer(std::string_view dataset, bool is_batch, Idx batch_size, Idx n_components, char const** components,
+               Idx const* elements_per_scenario, Idx const** indptrs, void const** data)
+        : dataset_{&meta_data().get_dataset(dataset)},
+          is_batch_{is_batch},
+          batch_size_{batch_size},
+          n_components_{n_components},
+          packer_{msgpack_buffer_} {
+        if (!is_batch_ && (batch_size_ != 1)) {
+            throw SerializationError{"For non-batch dataset, batch size should be one!\n"};
+        }
+        store_buffers(components, elements_per_scenario, indptrs, data);
+    }
+
+    void serialize(bool use_compact_list) {
```

Comment:
as a user, i would expect `serialize` to return a value (like in C# or JavaScript). Since the class is already named `Serializer`, I expect this function to be called in the constructor, or otherwise in the `get`ters. maybe this function should be private?

i understand that you want to give the user the option to `use_compact_list` but in that case, i guess the `get_msgpack` and `get_json` functions should expose that option

Reply:
done